### PR TITLE
Use cryptography and pyOpenSSL instead of M2Crypto

### LINKED ISSUE #411: fedmsg-0.18.3: tests fail in crypto
Trying to build fedmsg in rawhide fails in tests. 

```
Does fedmsg.config._recursive_update produce the expected result? ... ok
test_send_message (fedmsg.tests.test_core.TestCore)
send_message is deprecated ... ok
test_invalid_validator (fedmsg.tests.test_crypto_switching.TestCryptoSwitching)
Try to verify an x509 message when only gpg is allowed. ... ok
test_failed_validation (fedmsg.tests.test_crypto_x509.TestCryptoX509)
Try to fail validation. ... /var/tmp/rpm-tmp.FX8uYf: line 31: 10522 Segmentation fault      (core dumped) PYTHONPATH=$(pwd) python setup.py test
RPM build errors:
```

dmesg on the builder has: 

[Sat May 27 18:32:29 2017] python[10522]: segfault at 49 ip 00000000f6e09585 sp 00000000ffd29d00 error 4 in libcrypto.so.1.1.0e[f6c85000+22f000]

Perhaps we should move to x509_ng which doesn't use mcrypto, but python-crypto

### PR DESCRIPTION
**Pull request by jeremycline**
M2Crypto is Python 2 only and is in maintenance mode. Although pyOpenSSL
is _also_ in maintenance mode, it supports Python 3 and relies on
cryptography for its OpenSSL bindings. Currently the only thing
cryptography is missing (that pyOpenSSL has) is certificate chain
validation. Once that's available in cryptography, we can drop
pyOpenSSL.

Unfortunately, we still need to support M2Crypto because the current
versions of cryptography and pyOpenSSL in EL7 is below 1.6 and 16.1.0
respectively.

This patch uses cryptography and pyOpenSSL if available and falls back
to M2Crypto.

fixes #411

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

### CODE REVIEW

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py]: True *if* the policy…

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py:29]: Haha

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py]: This might make sense at warning level.

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py]: I recommend documenting the arguments and the return type.

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py:89]: Maybe put a timeout on this call to ```get()```.

File: fedmsg/crypto/utils.py
Comment (bowlofeggs) [inline fedmsg/crypto/utils.py:111]: Note that this could also raise an IOError, but that's probably OK.

File: fedmsg/crypto/x509.py
Comment (bowlofeggs) [inline fedmsg/crypto/x509.py:258]: Oh I see - you just moved this code block. I'd still recommend that timeout though… (maybe a separate PR or commit if you prefer - up to you).

File: fedmsg/crypto/x509_ng.py
Comment (bowlofeggs) [inline fedmsg/crypto/x509_ng.py:180]: I presume this step will catch expired certificates and non-properly signed certificates?

File: fedmsg/tests/crypto/test_x509.py
Comment (bowlofeggs) [inline fedmsg/tests/crypto/test_x509.py]: 2017?

File: fedmsg/tests/crypto/test_x509.py
Comment (bowlofeggs) [inline fedmsg/tests/crypto/test_x509.py:157]: How about a ```test_signed_by_expired_cert``` and ```test_signed_by_expired_ca``` (bad name, but what I mean is the signing cert is signed by an expired CA) as well?

File: fedmsg/tests/crypto/test_x509.py
Comment (bowlofeggs) [inline fedmsg/tests/crypto/test_x509.py:49]: I like this pattern of making a superclass to get the same tests run with the different implementations. Nice!

File: fedmsg/tests/crypto/test_x509.py
Comment (bowlofeggs) [inline fedmsg/tests/crypto/test_x509.py:226]: Do these tests work offline? If not, I see that you have a ```fedmsg/tests/test_certs/keys/crl.pem``` that you might could use instead…

**Review by bowlofeggs — APPROVED** [review]
This looks reasonable to me. Note however that I lack familiarity with the cryptography and pyopenssl libraries, so I am not able to comment on whether they were used correctly or not. Thus, I recommend getting a second opinion from someone who is familiar with those APIs before merging. On the other hand, I did like the tests you wrote and they helped give me confidence in the code ☺

File: fedmsg/crypto/x509.py
Comment (jeremycline) [reply fedmsg/crypto/x509.py:258]: I was actually thinking of doing a separate PR that changes this API a lot (one reason I didn't bother documenting it when I moved it). It does quite a few surprising things which I will probably address in the fix for #420.

File: fedmsg/crypto/x509_ng.py
Comment (jeremycline) [reply fedmsg/crypto/x509_ng.py:180]: Yup, it'll raise for that as well as revoked certificates and expired CRLs

File: fedmsg/tests/crypto/test_x509.py
Comment (jeremycline) [reply fedmsg/tests/crypto/test_x509.py:226]: They do, but only because I set the cache expiry to be ~47 years. There's no way to have the cert not downloaded from the Internet eventually and it defaults to the Fedora CA and CRL so even if you don't provide a URL it'll happen (all of which I want to change, I just haven't decided how yet).

File: fedmsg/tests/crypto/test_x509.py
Comment (jeremycline) [reply fedmsg/tests/crypto/test_x509.py:157]: Good idea! I have made it so, Number One.

File: fedmsg/crypto/utils.py
Comment (puiterwijk) [inline fedmsg/crypto/utils.py:60]: Maybe start the name with _ then?

File: fedmsg/crypto/x509.py
Comment (puiterwijk) [inline fedmsg/crypto/x509.py:49]: Perhaps add parentheses to indicate the and/or relations?

File: fedmsg/crypto/x509_ng.py
Comment (puiterwijk) [inline fedmsg/crypto/x509_ng.py:85]: While it's fine right now, it would be great if we could start migrating to new hashes, like sha256/sha512.
Maybe add an indicator which is used before the next version, so we can switch easier in the future?

File: fedmsg/crypto/x509_ng.py
Comment (puiterwijk) [inline fedmsg/crypto/x509_ng.py:200]: You should not use the common name ideally, this is being deprecated by the CA/Browser Forum baseline requirements.
While we are not reliant on those, I think ideally we should follow their move to the Subject Alternative Name.
This does mean there may be multiple records, and we might need to check all SAN values in the authorization step.
For the time being, for compatibility, we might add the Common Name to the list of SANs for checking purposes though and issue a warning if the CN was not in any SAN values.

File: fedmsg/crypto/x509_ng.py
Comment (puiterwijk) [inline fedmsg/crypto/x509_ng.py:204]: An idea would be a return any(...) over all the SANs.

File: fedmsg/tests/crypto/test_x509.py
Comment (puiterwijk) [inline fedmsg/tests/crypto/test_x509.py:58]: Ideally we would not depend on this website... Maybe use a file:// ?

**Review by puiterwijk — APPROVED** [review]
Overall this looks good, with some future suggestions.

File: fedmsg/crypto/x509_ng.py
Comment (jeremycline) [reply fedmsg/crypto/x509_ng.py:85]: I considered this, but I didn't want to make that change alongside all the other stuff that's already happening. I thought I filed an issue, but it seems like I hadn't, so I filed #438

Comment (puiterwijk) [reply fedmsg/crypto/x509_ng.py:85]: That makes perfect sense. This was just an idea now that you got me rereading the code :).

File: fedmsg/crypto/x509_ng.py
Comment (jeremycline) [reply fedmsg/crypto/x509_ng.py:200]: Yeah, I wanted to do this, but the M2Crypto version is only using the CN. I will definitely look into moving both of them to SAN (#439).

File: fedmsg/tests/crypto/test_x509.py
Comment (jeremycline) [reply fedmsg/tests/crypto/test_x509.py:58]: Most unfortunately, the way the code currently works it doesn't support file://. I set the cache expiry to be sometime in the mid 2030s so it should never actually try to fetch from that URL. I plan to re-write all the cert fetching code in the near future.

File: fedmsg/crypto/x509_ng.py
Comment (puiterwijk) [reply fedmsg/crypto/x509_ng.py:200]: The concrete sentence about Common Name from the CA/Browser Baselines requirements v1.4.5 (which is currently the latest stable, source https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.4.5.pdf):

````
Certificate Field: subject:commonName (OID 2.5.4.3)   
Required/Optional: Deprecated (Discouraged, but not prohibited) 
````

**Pull request merged**
